README.md
# PHP Prepended files

Dans **PHP**, il est possible d'automatiser l'intégration systématique de fichiers avant l'exécution et après l'exécution d'un script PHP. Ce sont respectivement les deux directives internes **auto_prepend_file** et **auto_append_file** qui permettent cette automatisation.

> Ces deux directives sont disponibles quelle que soit la version de PHP.

Pour utiliser ces fonctionnalités, il faut d'abord les configurer dans le fichier de configuration **PHP** ``php.ini``.

La particularité de ces deux fonctionnalités est qu'il faut spécifier, chaque fichier à inclure.

Pour simplifier l'utilisation de ces deux options, j'ai conçu un script qui parcourt les différents fichiers et dossiers qui se trouvent à son emplacement. Ainsi seul ce script doit être **prepended** ou **appended** dans le fichier de configuration `php.ini`.

Il a l'avantage également d'être configurable pour ajuster au mieux le comportement du script selon les différents sites virtuels qui pourraient être hébergés sur le serveur en évitant ainsi des erreurs de redéclaration de classes, de fonctions ou encore des fichiers de configuration pouvant provoquer des comportements non désirés.

Un exemple d'utilisation de l'`auto_prepend_file` est l'intégration de fonctions globales basiques propriétaires qui pourront être utilisées dans n'importe quel script PHP sans devoir à chaque fois faire l'inclusion manuellement.


## 1. Configuration du fichier ``php.ini``

> La configuration suivante est déjà effective dans la machine virtuelle **MCOSDE**.

Ci-dessous, la définition des deux directives issues du **manuel PHP** :

```
auto_prepend_file string
    Spécifie le nom d'un fichier qui sera automatiquement parcouru avant le fichier principal. 
    Ce fichier est inclus comme s'il l'avait été avec la fonction require, donc include_path est utilisé.
    
    La valeur spéciale none désactive l'ajout automatique.
```

```
auto_append_file string
    Spécifie le nom du fichier qui sera automatiquement parcouru après le fichier principal.
    Ce fichier est inclus comme s'il l'avait été avec la fonction require, donc include_path est utilisé.
    
    La valeur spéciale none désactive l'ajout automatique.
    
    Note: Si le script se termine par la fonction exit(), l'ajout automatique ne se fera pas.
```

Dans notre configuration, nous allons spécifier un fichier unique qui est ``auto_prepend_files.php``.

Nous allons donc éditer le fichier de configurtion ``php.ini`` à l'aide de l'éditeur ``nano`` :

```bash
sudo nano /etc/php5/fpm/php.ini
```

Cherchez la directive de configuration ``auto_prepend_file`` à l'aide de la combinaison de touches suivantes : `CTRL + W` puis tapez `auto_prepend_file`.

* Si la ligne commence par un **point-virgule** (`;`), supprimez-le (ligne commentée).
* Inscrivez le nom du script ``auto_prepend_files.php`` derrière le signe `=`.

Vous devriez avoir quelque chose comme ceci :

```
; Automatically add files before PHP document.
; http://php.net/auto-prepend-file
auto_prepend_file = auto_prepend_files.php
```

Nous n'avons pas spécifié l'emplacement complet vers ce fichier, car nous allons exploiter la fonctionnalité des ``include_path``. Lorsqu'on appelle un fichier, **PHP** cherche ce fichier dans une liste de dossier définis dans l'option `include_path`.

De nouveau dans l'éditeur nano, faites ``CTRL + W`` et tapez `include_path`.

* Si la ligne commence par un **point-virgule** (`;`), supprimez-le (ligne commentée).
* Pour séparer plusieurs valeur, utilisez les **deux-points** (`:`).
* Inscrivez l'emplacement du dossier dans lequel se trouve le script ``auto_prepend_files.php``, dans notre cas : `/var/www/prepends`.

Vous devriez avoir quelque chose comme ceci :

```
; UNIX : "/path1:/path2
include_path = ".:/usr/share/php:/var/www/prepends"
```

Répétez l'opération de configuration pour le **SAPI** ``CLI`` dont le fichier se trouve dans l'emplacement suivant :
``/etc/php5/cli/php.ini``

La configuration étant désormais terminée, il ne reste plus qu'à redémarrer le service **PHP** pour qu'elle soit effective :

```bash
sudo service php5-fpm restart
```

**Important** : Dans l'environnement de développement **[MCOSDE](https://gitlab-gre.viseo.net/MCOScheduler/MCOSDE)**, le dossier ``/var/www/prepends`` est un lien
symbolique pointant vers ``/home/mco/prepends``. Si vous avez une configuration différente, il faudra faire les
ajustements.


## 2. Configuration de ``auto_prepend_files.php`` :

Automatiser l'inclusion de nombreux fichiers présent dans des dossiers et sous-dossiers peut être nuisible lorsque plusieurs sites virtuels cohabitent sur le même serveur. Il fallait donc un système pour ajuster le comportement du script afin de lui préciser des règles d'exclusion d'intégration. Ce fichier de configuration est le fichier ``.ignore`` qui doit se trouver dans le même emplacement que le script `auto_prepend_files.php`.

Le principe de fonctionnement du fichier est très similaire aux fichiers de configurations de **NGINX**.

### 2.1. Les commentaires :

#### 2.1.1. Les commentaires en ligne

J'ai choisi d'intégrer les trois modèles de commentaire en ligne les plus courants :

* ``#`` **Dièse** : Utilisé dans les fichiers de configuration **NGINX**, **MySQL**, **Bash**, **C/C++** et bien d'autre.
* ``;`` **Point-virgule** : Utilisé dans les fichiers de configuration **PHP**.
* ``//`` **Double slash** : Utilisé dans les scripts **PHP**, **JavaScript**.

#### 2.1.2. Les commentaires multi-ligne

Bien qu'il soit possible de précéder chaque ligne du caractère de commentaire ***inline***, pour les gros pavés de texte, le commentaire multi-ligne reste plus approprié. Là encore j'ai utilisé le modèle le plus courant qui est le suivant.

```
/* Comment Block on one line */

/**
 * Multi-ligne
 * Comment
 **/
 
/**
DIRTY
COMMENT
BLOCK
**/
```


### 2.2. Les entités :

Dans les univers **Unix** et **Linux**, tout est fichier, même les dossiers.
Fonctionnellement parlant, en revanche on distingue les deux entités :

* Les fichiers : ``FILES``
* Les dossiers : ``FOLDERS``

L'entité ``FILES`` sert à indiquer les fichiers à exclure.

L'entité ``FOLDERS`` sert à indiquer les dossiers à exclure.

Exclure un dossier comprend donc l'ensemble des fichiers et sous-dossier qu'il contient.


### 2.3. Les blocks d'instructions :

Tout comme **NGINX**, les instructions de configuration sont placées dans une structure constituant un **bloc**.
Un bloc se présente sous la forme suivante :

```
condition_src condition_value {
    (...)
    FILES {
        file_name.ext
    }
    (...)
}
```

Pour créer une règle d'exclusion, il faut créer un bloc muni d'une condition de réalisation puis de lister les valeurs à exclure pour la ou les entités désirées : ``FILES`` et/ou ``FOLDER``.


### 2.4. Les conditions

Le système de condition porte sur les données disponibles dans la variable super-globale ``$_SERVER`` disponible sous **Apache** (SAPI **apache**), **NGINX** (SAPI **FPM-FastCGI**) et en ligne de commande (SAPI **CLI**).

> SAPI : Serveur Application Programming Interface.

Dans la formule suivante : ``condition_src condition_value``, ``condition_src`` prendra la valeur d'une clé existante dans la variable super-globale ``$_SERVER``, tandis que ``condition_value`` sera la valeur à remplir pour traiter l'ensemble des instructions enfants à ce bloc d'instruction.


### 2.5. Expressions Réguliéres

L'ensemble des valeurs saisies, qu'il s'agisse d'une valeur de condition, d'un nom de fichier ou dossier à exclure, celles-ci sont traitées en tant que valeur exacte.
Il est possible de rendre ces valeurs plus souples en utilisant les expressions régulières. Pour transformer une valeur exacte en expression régulière, il suffit de la précéder à l'aide du caractère **tilde** ``~``.


## Annexe 1 - Server keys

Ci-dessous, la liste complète des clés admises pour définir une condition.

> La disponibilité des clés pour Apache n'ont pas été évaluées.

```
                                 +-----------------------+
                                 |          SAPI         |
+--------------------------------+-------+-------+-------+-------------------------------------------
|           NAME OF KEY          | APACH | NGINX |  CLI  | DESCRIPTION (Value sample)
+--------------------------------+-------+-------+-------+-------------------------------------------
|                           USER |       |   X   |   X   | www-data
+--------------------------------+-------+-------+-------+-------------------------------------------
|                           HOME |       |   X   |   X   | /var/www
+--------------------------------+-------+-------+-------+-------------------------------------------
|                      FCGI_ROLE |       |   X   |       | RESPONDER
+--------------------------------+-------+-------+-------+-------------------------------------------
|                      PATH_INFO |       |   X   |       |
+--------------------------------+-------+-------+-------+-------------------------------------------
|                SCRIPT_FILENAME |       |   X   |   X   | /var/www/browse.php
+--------------------------------+-------+-------+-------+-------------------------------------------
|                   QUERY_STRING |       |   X   |       |
+--------------------------------+-------+-------+-------+-------------------------------------------
|                 REQUEST_METHOD |       |   X   |       | GET
+--------------------------------+-------+-------+-------+-------------------------------------------
|                   CONTENT_TYPE |       |   X   |       |
+--------------------------------+-------+-------+-------+-------------------------------------------
|                 CONTENT_LENGTH |       |   X   |       |
+--------------------------------+-------+-------+-------+-------------------------------------------
|                    SCRIPT_NAME |       |   X   |   X   | /browse.php
+--------------------------------+-------+-------+-------+-------------------------------------------
|                    REQUEST_URI |       |   X   |       | /PHP.Classes/SYSLang.V.1.0/
+--------------------------------+-------+-------+-------+-------------------------------------------
|                   DOCUMENT_URI |       |   X   |       | /browse.php
+--------------------------------+-------+-------+-------+-------------------------------------------
|                  DOCUMENT_ROOT |       |   X   |   X   | /var/www
+--------------------------------+-------+-------+-------+-------------------------------------------
|                SERVER_PROTOCOL |       |   X   |       | HTTP/1.1
+--------------------------------+-------+-------+-------+-------------------------------------------
|              GATEWAY_INTERFACE |       |   X   |       | CGI/1.1
+--------------------------------+-------+-------+-------+-------------------------------------------
|                SERVER_SOFTWARE |       |   X   |       | nginx/1.6.2
+--------------------------------+-------+-------+-------+-------------------------------------------
|                    REMOTE_ADDR |       |   X   |       | 74.123.43.22
+--------------------------------+-------+-------+-------+-------------------------------------------
|                    REMOTE_PORT |       |   X   |       | 53211
+--------------------------------+-------+-------+-------+-------------------------------------------
|                    SERVER_ADDR |       |   X   |       | 192.168.1.253
+--------------------------------+-------+-------+-------+-------------------------------------------
|                    SERVER_PORT |       |   X   |       | 80
+--------------------------------+-------+-------+-------+-------------------------------------------
|                    SERVER_NAME |       |   X   |       | neoblaster.fr
+--------------------------------+-------+-------+-------+-------------------------------------------
|                REDIRECT_STATUS |       |   X   |       | 200
+--------------------------------+-------+-------+-------+-------------------------------------------
|                            CWD |       |   X   |       | Custom server param sent by FAST_PARAM
+--------------------------------+-------+-------+-------+-------------------------------------------
|                      HTTP_HOST |       |   X   |       | neoblaster.fr
+--------------------------------+-------+-------+-------+-------------------------------------------
|                HTTP_CONNECTION |       |   X   |       | keep-alive
+--------------------------------+-------+-------+-------+-------------------------------------------
| HTTP_UPGRADE_INSECURE_REQUESTS |       |   X   |       | 1
+--------------------------------+-------+-------+-------+-------------------------------------------
|                HTTP_USER_AGENT |       |   X   |       | Mozilla/5.0 (Windows NT 6.1; WOW64) (...)
+--------------------------------+-------+-------+-------+-------------------------------------------
|                    HTTP_ACCEPT |       |   X   |       |
+--------------------------------+-------+-------+-------+-------------------------------------------
|                       HTTP_DNT |       |   X   |       | 1
+--------------------------------+-------+-------+-------+-------------------------------------------
|           HTTP_ACCEPT_ENCODING |       |   X   |       | gzip, deflate, sdch
+--------------------------------+-------+-------+-------+-------------------------------------------
|           HTTP_ACCEPT_LANGUAGE |       |   X   |       | fr-FR,fr;q=0.8,en-US;q=0.6,en;q=0.4
+--------------------------------+-------+-------+-------+-------------------------------------------
|                    HTTP_COOKIE |       |   X   |       | SYSLang_LANG=fr-FR
+--------------------------------+-------+-------+-------+-------------------------------------------
|                       PHP_SELF |       |   X   |   X   | /browse.php
+--------------------------------+-------+-------+-------+-------------------------------------------
|             REQUEST_TIME_FLOAT |       |   X   |   X   | 1485468784.8032
+--------------------------------+-------+-------+-------+-------------------------------------------
|                   REQUEST_TIME |       |   X   |   X   | 1485468784
+--------------------------------+-------+-------+-------+-------------------------------------------
|                 XDG_SESSION_ID |       |       |   X   | c2
+--------------------------------+-------+-------+-------+-------------------------------------------
|                           TERM |       |       |   X   | xterm
+--------------------------------+-------+-------+-------+-------------------------------------------
|                          SHELL |       |       |   X   | /bin/bash
+--------------------------------+-------+-------+-------+-------------------------------------------
|                     SSH_CLIENT |       |       |   X   | 74.123.43.22 50163 22
+--------------------------------+-------+-------+-------+-------------------------------------------
|                        SSH_TTY |       |       |   X   | /dev/pts/0
+--------------------------------+-------+-------+-------+-------------------------------------------
|                      LS_COLORS |       |       |   X   | rs=0:di=01;34:ln=01;36:mh=00:pi=40; (...)
+--------------------------------+-------+-------+-------+-------------------------------------------
|                           MAIL |       |       |   X   | /var/mail/pi
+--------------------------------+-------+-------+-------+-------------------------------------------
|                           PATH |       |       |   X   | /usr/local/sbin:/usr/local/bin:/usr/sbin
+--------------------------------+-------+-------+-------+-------------------------------------------
|                            PWD |       |       |   X   | /var/www/Prepends
+--------------------------------+-------+-------+-------+-------------------------------------------
|                           LANG |       |       |   X   | en_GB.UTF-8
+--------------------------------+-------+-------+-------+-------------------------------------------
|                          SHLVL |       |       |   X   | 1
+--------------------------------+-------+-------+-------+-------------------------------------------
|                        LOGNAME |       |       |   X   | pi
+--------------------------------+-------+-------+-------+-------------------------------------------
|                 SSH_CONNECTION |       |       |   X   | 74.123.43.22 50163 192.168.1.253 22
+--------------------------------+-------+-------+-------+-------------------------------------------
|                XDG_RUNTIME_DIR |       |       |   X   | /run/user/1000
+--------------------------------+-------+-------+-------+-------------------------------------------
|                         OLDPWD |       |       |   X   | /var/www
+--------------------------------+-------+-------+-------+-------------------------------------------
|                              _ |       |       |   X   | /usr/bin/php
+--------------------------------+-------+-------+-------+-------------------------------------------
|                PATH_TRANSLATED |       |       |   X   | auto_prepend_files.php
+--------------------------------+-------+-------+-------+-------------------------------------------
|                           argv |       |       |   X   | ARRAY
+--------------------------------+-------+-------+-------+-------------------------------------------
|                           argc |       |       |   X   | 1
+--------------------------------+-------+-------+-------+-------------------------------------------
```

CHANGELOG.md
## 1.3.0 (2017-11-26)

- [FIXED] Correction de la fonction `setup` qui utilisait le `REQUEST_URI` sans vérification d'existante qui
échouait dans le cas de ré-écriture d'URL.
- [UPDATE] Mise à jour du fichier de configuration `.ignore` pour ignorer la déclaration de constante pour le
**SAPI CLI**.
- [ADDED] Création de la function `array_rmap($callback, &$arr)` permettant l'opération sur les tableaux sans
qu'il soit nécessaire de récupérer le résultat.


## 1.2.1 (2017-10-02)

- [FIXED] Mise à jour du fichier de configuration `.ignore` pour ignorer tous les fichiers **Markdown**.


## 1.2.0 (2017-10-02)

- [FIXED] Correction du setup `setup.const.document_root.php` generant une erreur `E_NOTICE` pour `CLI`.


## 1.1.0 (2017-10-02)

- [CHANGED] Modification du nom du dossier d'inclusion ``Prepends`` par `prepends` afin de s'accorder avec le projet
**[MCOSDE](https://gitlab-gre.viseo.net/MCOScheduler/MCOSDE)**




[!ADDED]:#
[!FIXED]:#
[!CHANGED]:#
[!REMOVED]:#
[!SECURITY]:#
[!DEPRECATED]:#
[!OTER]:#
[!BUGFIX]:#


auto_prepend_files.php
<?php
/**
 * auto_prepend_files.php
 *
 * Par le biais de ce script auto prepended depuis la configuration de `php.ini`, celui-ci permet de charger
 * autant de fichiers que nécessaire qui sont déposés dans le dossier courant, sans qu'il soit nécessaire
 * de les ajouter dans la directive `auto_prepend_file` du fichier `php.ini`.
 *
 * Description Fonctionnelle :
 *
 * 1. Création d'une variable unique $_PREPEND pour effacer toutes traces d'execution du présent script.
 *
 * 2. Traitement
 *  * 2.1. S'ajouter en tant que fichier exclus
 *  * 2.2. Si un fichier .ignore existe, on le parse
 *  * 2.3. On parcours tous les dossiers/fichiers dans l'emplacement courant
 *  * 2.4. Effacer toutes les traces (variables & fichier temporaire)
 *
 * @author    Nicolas DUPRE
 * @release   19/09/2017
 * @version   1.0.0
 * @package   Index
 */


/**
 * @var array $_PREPEND Environnement de travail du script afin de laisser pur la global $GLOBALS
 */
$_PREPEND = [
    /**
     * boolean Active le mode debuggage permettant l'affichage des sorties
     */
    "debug" => false,

    /**
     * array Règles d'exclusions
     */
    "ignore" => [
        /**
         * array Liste des dossiers à exclure
         */
        "folders" => [],
        "folders_regexp" => [],
        /**
         * array Liste des fichiers à exclure
         */
        "files" => [],
        "files_regexp" => []
    ]
];


/**
 * Affiche le message demandé dans le flux de sortie
 */
function apf_stdout($message, $stop = false){
    global $_PREPEND;

    $_PREPEND["output"] = fopen("php://output", "r");

    if(@$_PREPEND["debug"]){
        fputs($_PREPEND["output"], $message);

        if($stop) exit;
    }
}

/**
 * Lit de manière récursive le dossier indiqué et inclus les fichiers trouvé
 */
function apf_read_folder($full_path){
    global $_PREPEND;
    $folders = scandir($full_path);

    /** Parcourir le résultat obtenu */
    foreach ($folders as $fkey => $fvalue){
        /** Si l'entrée ne commence pas par un point */
        if(!preg_match("#^\.#", $fvalue)){
            apf_stdout("--> CHECKING FILE $fvalue" . PHP_EOL);

            /** S'il s'agit d'un dossier */
            if(is_dir("$full_path/$fvalue")){
                /** S'il n'est pas explicitement exclus */
                if(!in_array($fvalue, $_PREPEND["ignore"]["folders"])){
                    $allowed = true;

                    /** S'il n'est pas exclus par RegExp */
                    foreach ($_PREPEND["ignore"]["folders_regexp"] as $exclude_pattern){
                        if(preg_match("#$exclude_pattern#", $fvalue)){
                            $allowed = false;
                            break;
                        }

                    }

                    if($allowed) apf_read_folder("$full_path/$fvalue");
                }
            }
            /** S'il s'agit d'un fichier */
            else {
                /** S'il n'est pas explicitement exclus */
                if(!in_array($fvalue, $_PREPEND["ignore"]["files"])){
                    $allowed = true;

                    /** S'il n'est pas exclus par RegExp */
                    foreach ($_PREPEND["ignore"]["files_regexp"] as $exclude_pattern){
                        if(preg_match("#$exclude_pattern#", $fvalue)){
                            $allowed = false;
                            break;
                        }
                    }

                    if($allowed) require_once "$full_path/$fvalue";
                }
            }
        }
    }
}


/**
 * Activation du niveau d'erreur à "afficher tout"
 */
if($_PREPEND["debug"]) error_reporting(E_ALL);


/**
 * S'auto-exclure des fichiers
 *
 * Permet tout renommage sans modification du script
 * Empêche de créer une boucle infinie
 */
$_PREPEND["ignore"]["files"][] = basename(__FILE__);


/**
 * Processing
 */
apf_stdout("<pre>");

// Traitement du fichier .ignore
if(file_exists(__DIR__ . "/.ignore")){
    apf_stdout("--> .ignore FILE PROCESSING" . PHP_EOL);

    if(!file_exists(__DIR__ . "/.tmp")) mkdir(__DIR__ . "/.tmp", 0775);

    $_PREPEND["skip_file_cleansing"] = false;
    $_PREPEND["ignore_file"] = fopen(__DIR__ . "/.ignore", "r");
    $_PREPEND["tmp_file_name"] = md5(file_get_contents(__DIR__ . "/.ignore"));

    /**
     * Inutile de retraiter un fichier qui ne change que rarement
     * Evite les conflits d'écriture en cas de traitement simultané
     * Permet la sauvegarde des ancienne configuration
     */
    if(file_exists(__DIR__ . "/.tmp/" . $_PREPEND["tmp_file_name"])) $_PREPEND["skip_file_cleansing"] = true;

    /**
     * Nettoyage du fichier .ignore
     */
    if(!$_PREPEND["skip_file_cleansing"]){
        /** Création d'un fichier temporaire de destination */
        $_PREPEND["ignore_tmp_file"] = fopen(__DIR__ . "/.tmp/" . $_PREPEND["tmp_file_name"], "w+");

        /** Rechercher le BREAK_SIGNAL dans le fichier d'origine */
        while($_PREPEND["buffer"] = fgets($_PREPEND["ignore_file"])){
            if(preg_match("/#\s*BREAK_SIGNAL/", $_PREPEND["buffer"])){
               fclose($_PREPEND["ignore_tmp_file"]);
               break;
            } else {
                fputs($_PREPEND["ignore_tmp_file"], $_PREPEND["buffer"]);
            }
        }
        fclose($_PREPEND["ignore_file"]);

        /** Suppression des commentaires */
        $_PREPEND["ignore_cfg"] = file_get_contents(__DIR__ . "/.tmp/" . $_PREPEND["tmp_file_name"]);

        apf_stdout("--> FILE BEFORE COMMENTS CLEANSING" . PHP_EOL);
        apf_stdout($_PREPEND["ignore_cfg"]);

        $_PREPEND["ignore_cfg"] = preg_replace("/^\s*(#|;).*/mi", "", $_PREPEND["ignore_cfg"]);
        $_PREPEND["ignore_cfg"] = preg_replace("#//.*#mi", "", $_PREPEND["ignore_cfg"]);
        $_PREPEND["ignore_cfg"] = preg_replace("#\/+\*+(.|\s)*\*+\/+#miU", "", $_PREPEND["ignore_cfg"]);

        apf_stdout("--> FILE AFTER COMMENTS CLEANSING" . PHP_EOL);
        apf_stdout($_PREPEND["ignore_cfg"]);

        file_put_contents(__DIR__ . "/.tmp/" . $_PREPEND["tmp_file_name"], $_PREPEND["ignore_cfg"]);
    }

    /** Lecture des règles */
    $_PREPEND["ignore_file_light"] = fopen(__DIR__ . "/.tmp/" . $_PREPEND["tmp_file_name"], "r");
    $_PREPEND["current_level"] = 0;
    $_PREPEND["accept_level"] = Array(true); // Niveau 0 est otujours admis, car vaut pour le wildcard *
    $_PREPEND["in_folders_block"] = false;
    $_PREPEND["in_files_block"] = false;

    while($_PREPEND["buffer"] = fgets($_PREPEND["ignore_file_light"])){
        /** Supprimer le retour chariot */
        $_PREPEND["buffer"] = str_replace("\n", "", $_PREPEND["buffer"]);

        /** RAZ du flag indiquant de ne pas traiter cette ligne de buffer car vaut pour instruction */
        $_PREPEND["skip_buffer"] = false;

        /** Si la ligne n'est pas vide */
        if(!preg_match("#^\s*$#", $_PREPEND["buffer"])){
            apf_stdout("BUFFER :: #$_PREPEND[buffer]#" . PHP_EOL);

            /** Vérifier s'il s'agit d'un groupe FILES */
            if(preg_match("#^\s*FILES\s*\{#i", $_PREPEND["buffer"]) && $_PREPEND["accept_level"][$_PREPEND["current_level"]]){
                apf_stdout("--> BLOCK FILE FOUND & LEVEL ALLOW;".PHP_EOL);
                $_PREPEND["in_files_block"] = true;
                $_PREPEND["skip_buffer"] = true;
            }

            /** Vérifier s'il s'agit d'un groupe FOLDER et que celui-ci est approuvé dans son niveau d'imbrication */
            if(preg_match("#^\s*FOLDERS\s*\{#i", $_PREPEND["buffer"])){
                apf_stdout("--> BLOCK FOLDER FOUND & LEVEL ALLOW;".PHP_EOL);
                $_PREPEND["in_folders_block"] = true;
                $_PREPEND["skip_buffer"] = true;
            }

            /** Vérifier s'il s'agit d'une règle d'exclusion (instruction) */
            if(preg_match("#^\s*([a-zA-Z_]+)\s+([a-zA-Z0-9-_.\s\/\(\);,=:\*^$?~]+)\s+\{#", $_PREPEND["buffer"], $_PREPEND["matches"])){
                apf_stdout("--> INSTRUCTION FOUND;" . PHP_EOL);

                /** On entre dans un niveau/sous/niveau */
                apf_stdout("--> INCREASE LEVEL FROM $_PREPEND[current_level]");
                $_PREPEND["current_level"]++;
                apf_stdout(" to $_PREPEND[current_level]" . PHP_EOL);

                /** Translation de variable pour "simplifier" */
                $_PREPEND["key"] = $_PREPEND["matches"][1];
                $_PREPEND["value"] = $_PREPEND["matches"][2];

                if(array_key_exists($_PREPEND["key"], $_SERVER)){
                    apf_stdout("--> KEY EXIST IN &amp;_SERVER;" . PHP_EOL);

                    /** Si la valeur commence par un tilde, alors c'est une RegExp */
                    if(preg_match("#^~#", $_PREPEND["value"])){
                        apf_stdout("--> VALUE READ AS REGEXP;" . PHP_EOL);
                        /** Suppression du tilde */
                        $_PREPEND["value"] = preg_replace("#^~#", "", $_PREPEND["value"], 1);

                        if(preg_match("#$_PREPEND[value]#", $_SERVER[$_PREPEND["key"]])){
                            apf_stdout("--> &amp;_SERVER VALUE MATCH WITH REGEXP PATTERN;");
                            $_PREPEND["accept_level"][$_PREPEND["current_level"]] = true;
                        } else {
                            apf_stdout("--> &amp;_SERVER VALUE DOES NOT MATCH WITH REGEXP PATTERN;");
                            $_PREPEND["accept_level"][$_PREPEND["current_level"]] = false;
                        }
                    }
                    /** Sinon si la valeur est exactement celle attendue */
                    else if ($_SERVER[$_PREPEND["key"]] == $_PREPEND["value"]) {
                        apf_stdout("--> VALUE READ AS STRING;" . PHP_EOL);
                        apf_stdout("--> VALUE EXACTLY EQUAL;" . PHP_EOL);
                        $_PREPEND["accept_level"][$_PREPEND["current_level"]] = true;
                    }
                    /** Sinon, ignorer le block */
                    else {
                        $_PREPEND["accept_level"][$_PREPEND["current_level"]] = false;
                    }

                } else {
                    $_PREPEND["accept_level"][$_PREPEND["current_level"]] = false;
                }

                $_PREPEND["skip_buffer"] = true;
            }

            /** Vérifier s'il s'agit d'une fermeture de block */
            if(preg_match("#^\s*\}\s*$#", $_PREPEND["buffer"])){
                apf_stdout("--> BLOCK END FOUND - CLOSE LEVEL;" . PHP_EOL);

                /** Gestion de niveau s'il ne s'agisssait pas d'un block FOLDERS ou FILES */
                if(!$_PREPEND["in_files_block"] && !$_PREPEND["in_folders_block"]){
                    $_PREPEND["accept_level"] = false;
                    $_PREPEND["current_level"]--;
                }

                /** Si nous étions dans un block FILES ou FOLDERS, il est cloturé, on n'enregistre plus */
                if($_PREPEND["in_files_block"]) $_PREPEND["in_files_block"] = false;
                if($_PREPEND["in_folders_block"]) $_PREPEND["in_folders_block"] = false;

                $_PREPEND["skip_buffer"] = true;
            }

            /** Traitement du buffer */
            if(!$_PREPEND["skip_buffer"]){
                apf_stdout("--> TEXT LINE, PROCESS;" . PHP_EOL);

                preg_match("#^\s*(.*)\s*$#i", $_PREPEND["buffer"], $_PREPEND["matches"]);

                /** Permet d'ignorer les lignes vide entre deux valeurs au sein du même block */
                if(count($_PREPEND["matches"] >= 2)){
                    apf_stdout("--> CHECK FOR EXCLUSION;" . PHP_EOL);

                    $regexp_store = "";

                    if(preg_match("#^~#", $_PREPEND["matches"][1])){
                        $regexp_store = "_regexp";
                        $_PREPEND["matches"][1] = preg_replace("#^~#", "", $_PREPEND["matches"][1]);
                    }

                    if($_PREPEND["in_files_block"]){
                        apf_stdout("--> LINE ADD TO FILES EXCLUSION;" . PHP_EOL);
                        $_PREPEND["ignore"]["files$regexp_store"][] = $_PREPEND["matches"][1];
                    } else if($_PREPEND["in_folders_block"]) {
                        apf_stdout("--> LINE ADD TO FOLDERS EXCLUSION;" . PHP_EOL);
                        $_PREPEND["ignore"]["folders$regexp_store"][] = $_PREPEND["matches"][1];
                    } else {
                        apf_stdout("--> LINE SKIPPED;" . PHP_EOL);
                    }
                }
            }

            apf_stdout(PHP_EOL . PHP_EOL);
        }

    }

    fclose($_PREPEND["ignore_file_light"]);
}

apf_stdout("IGNORE RULES RESULT : " . PHP_EOL);
apf_stdout(print_r($_PREPEND["ignore"], true));

apf_stdout("AUTO_PREPEND_FILES SCRIPT &amp;_PREPEND FOOTPRINT : " . PHP_EOL);
apf_stdout(print_r($GLOBALS["_PREPEND"], true));


// Lecture du dossier courant
apf_read_folder(__DIR__);


// Cleansing
@unlink(__DIR__ . "/.ignore_tmp");
unset($GLOBALS["_PREPEND"]);
unset($GLOBALS["regexp_store"]);

//print_r($GLOBALS);